x-pack/apm-server/sampling/eventstorage/storage_manager.go

```diff
@@ -53,288 +73,190 @@ type StorageManager struct {
+		sm.eventDB.Compact([]byte{lbPrefix}, []byte{ubPrefix}, false),
+		sm.decisionDB.Compact([]byte{lbPrefix}, []byte{ubPrefix}, false),
 	)
-	err := filepath.WalkDir(s.storageDir, func(path string, d fs.DirEntry, _ error) error {
-		if !rootVisited {
-			rootVisited = true
-			return nil
-		}
-		filename := filepath.Base(path)
-		if filename == subscriberPositionFile {
-			return nil
-		}
-		switch ext := filepath.Ext(filename); ext {
-		case ".sst":
-			sstFiles++
-		case ".vlog":
-			vlogFiles++
-		default:
-			otherFilenames = append(otherFilenames, filename)
-		}
-		return os.RemoveAll(path)
-	})
-	s.logger.Infof("deleted badger files: %d SST files, %d VLOG files, %d other files: [%s]",
-		sstFiles, vlogFiles, len(otherFilenames), strings.Join(otherFilenames, ", "))
-	return err
 }
 
-func (s *StorageManager) ReadSubscriberPosition() ([]byte, error) {
-	s.subscriberPosMu.Lock()
-	defer s.subscriberPosMu.Unlock()
-	return os.ReadFile(filepath.Join(s.storageDir, subscriberPositionFile))
+func (sm *StorageManager) ReadSubscriberPosition() ([]byte, error) {
+	sm.subscriberPosMu.Lock()
+	defer sm.subscriberPosMu.Unlock()
+	return os.ReadFile(filepath.Join(sm.storageDir, subscriberPositionFile))
 }
 
-func (s *StorageManager) WriteSubscriberPosition(data []byte) error {
-	s.subscriberPosMu.Lock()
-	defer s.subscriberPosMu.Unlock()
-	return os.WriteFile(filepath.Join(s.storageDir, subscriberPositionFile), data, 0644)
+func (sm *StorageManager) WriteSubscriberPosition(data []byte) error {
+	sm.subscriberPosMu.Lock()
+	defer sm.subscriberPosMu.Unlock()
+	return os.WriteFile(filepath.Join(sm.storageDir, subscriberPositionFile), data, 0644)
 }
 
-func (s *StorageManager) NewReadWriter() *ManagedReadWriter {
-	return &ManagedReadWriter{
-		sm: s,
-	}
-}
-
-// ManagedReadWriter is a read writer that is transparent to badger DB changes done by StorageManager.
-// It is a wrapper of the ShardedReadWriter under StorageManager.
-type ManagedReadWriter struct {
-	sm *StorageManager
-}
-
-func (s *ManagedReadWriter) ReadTraceEvents(traceID string, out *modelpb.Batch) error {
-	s.sm.mu.RLock()
-	defer s.sm.mu.RUnlock()
-	return s.sm.rw.ReadTraceEvents(traceID, out)
-}
-
-func (s *ManagedReadWriter) WriteTraceEvent(traceID, id string, event *modelpb.APMEvent, opts WriterOpts) error {
-	ok := s.sm.mu.TryRLock()
-	if !ok {
-		return errDropAndRecreateInProgress
-	}
-	defer s.sm.mu.RUnlock()
-	return s.sm.rw.WriteTraceEvent(traceID, id, event, opts)
+func (sm *StorageManager) NewReadWriter() StorageLimitReadWriter {
+	return NewStorageLimitReadWriter(sm, SplitReadWriter{
+		eventRW:    sm.eventStorage.NewReadWriter(),
+		decisionRW: sm.decisionStorage.NewReadWriter(),
+	})
 }
 
-func (s *ManagedReadWriter) WriteTraceSampled(traceID string, sampled bool, opts WriterOpts) error {
-	ok := s.sm.mu.TryRLock()
-	if !ok {
-		return errDropAndRecreateInProgress
+// NewBypassReadWriter returns a SplitReadWriter directly reading and writing to the database,
```

Comment:
This doesn't seem to really be needed. I just replaced all uses with NewReadWriter locally, and removed the Close calls (which are no-ops on SplitReadWriter anyway), and all the tests still pass.

Reply:
This was originally intended to view a lower level rw for minimal perf overhead and an accurate & independent view of db, but in a lot of usages (e.g. in processor_test.go), they really shouldn't care about the implementation. I've removed it to minimize the confusion. If we find the need to do some benchmarking closer to db, we can add it back later.